File: packages/es-markdown-core/src/markup-dumper.ts
import type { BlockQuote } from "./model/block-quote.js";
import type { CodeBlock } from "./model/code-block.js";
import type { Code } from "./model/code.js";
import type { DirectiveBlock } from "./model/directive-block.js";
import type { Directive } from "./model/directive.js";
import type { Document } from "./model/document.js";
import type { Emphasis } from "./model/emphasis.js";
import type { FootnoteDefinition, FootnoteReference } from "./model/footnote.js";
import type { FormulaBlock } from "./model/formula-block.js";
import type { Formula } from "./model/formula.js";
import type { Heading } from "./model/heading.js";
import type { HTMLBlock } from "./model/html-block.js";
import type { HTML } from "./model/html.js";
import type { Image } from "./model/image.js";
import type { Link } from "./model/link.js";
import type { List, ListItem } from "./model/list.js";
import type { Markup } from "./model/markup.js";
import type { Paragraph } from "./model/paragraph.js";
import type { Strikethrough } from "./model/strikethrough.js";
import type { Strong } from "./model/strong.js";
import type { Table, TableCell, TableRow } from "./model/table.js";
import type { Text } from "./model/text.js";
import type { Scope } from "./values.js";
import { visit, type MarkupVisitor } from "./markup-visitor.js";
import { MarkupWalker, WalkEvent } from "./markup-walker.js";

/** Produces the canonical diagnostic tree for immutable Markdown markup. */
export class MarkupDumper {
    private constructor() {}

    /**
     * Returns the canonical diagnostic dump for the subtree rooted at
     * `node` (the whole document by default), resolving absolute scopes
     * through the snapshot. A non-document subtree prints scopes with the
     * subtree as origin: the root's start line becomes line 1.
     * Position-free markers (0:0..0:0) print unchanged.
     */
    static dump(document: Document, node: Markup = document): string {
        const origin = document.scope(node).start.line;
        const offset = origin > 0 ? origin - 1 : 0;
        const remainingChildren: number[] = [];
        const lines: string[] = [];

        new MarkupWalker().walk(document, node, (event, current, scope) => {
            if (event === WalkEvent.entering) {
                const record = visit(current, dumpVisitor);
                const line = record.line(rebased(scope, offset));
                if (remainingChildren.length === 0) {
                    lines.push(line);
                } else {
                    const parent = remainingChildren.length - 1;
                    const prefix = remainingChildren
                        .slice(0, -1)
                        .map((remaining) => (remaining > 0 ? "│   " : "    "))
                        .join("");
                    const connector = remainingChildren[parent] === 1 ? "└── " : "├── ";
                    lines.push(prefix + connector + line);
                    remainingChildren[parent] = remainingChildren[parent]! - 1;
                }
                remainingChildren.push(record.children);
            } else {
                if (remainingChildren.pop() !== 0) throw new Error("walker exited before its children");
            }
        });
        return `${lines.join("\n")}\n`;
    }
}

interface PendingRecord {
    readonly line: (scope: string) => string;
    readonly children: number;
}

const dumpVisitor: MarkupVisitor<PendingRecord> = {
    visitDocument: (node: Document) => record("Document", [], node.content.length),
    visitBlockQuote: (node: BlockQuote) => record("BlockQuote", [], node.content.length),
    visitParagraph: (node: Paragraph) => record("Paragraph", [], node.content.length),
    visitHeading: (node: Heading) => record("Heading", [`level=${node.level}`], node.content.length),
    visitThematicBreak: () => record("ThematicBreak"),
    visitList: (node: List) =>
        record(
            "List",
            [`flavor=${node.flavor}`, `start=${node.start ?? "null"}`, `tight=${node.tight}`],
            node.items.length
        ),
    visitListItem: (node: ListItem) => record("ListItem", [`checked=${node.checked ?? "null"}`], node.content.length),
    visitCodeBlock: (node: CodeBlock) =>
        record("CodeBlock", [
            `mode=${node.mode}`,
            `info=${optionalString(node.info)}`,
            `language=${optionalString(node.language)}`,
            `literal=${jsonString(node.literal)}`,
            `fenced=${node.fenced}`,
            `closed=${node.closed}`
        ]),
    visitHTMLBlock: (node: HTMLBlock) => record("HTMLBlock", [`literal=${jsonString(node.literal)}`]),
    visitFormulaBlock: (node: FormulaBlock) =>
        record("FormulaBlock", [`mode=${node.mode}`, `literal=${jsonString(node.literal)}`]),
    visitTable: (node: Table) => record("Table", [`alignments=[${node.alignments.join(",")}]`], 1 + node.rows.length),
    visitTableRow: (node: TableRow) => record("TableRow", [`isHeader=${node.isHeader}`], node.cells.length),
    visitTableCell: (node: TableCell) => record("TableCell", [], node.content.length),
    visitDirectiveBlock: (node: DirectiveBlock) =>
        record(
            "DirectiveBlock",
            directiveFields(node.mode, node.name, node.attributes, node.label?.length ?? null),
            (node.label?.length ?? 0) + node.content.length
        ),
    visitFootnoteDefinition: (node: FootnoteDefinition) =>
        record("FootnoteDefinition", [`id=${jsonString(node.label)}`], node.content.length),
    visitText: (node: Text) => record("Text", [`literal=${jsonString(node.literal)}`]),
    visitSoftBreak: () => record("SoftBreak"),
    visitLineBreak: () => record("LineBreak"),
    visitCode: (node: Code) => record("Code", [`mode=${node.mode}`, `literal=${jsonString(node.literal)}`]),
    visitHTML: (node: HTML) => record("HTML", [`literal=${jsonString(node.literal)}`]),
    visitFormula: (node: Formula) => record("Formula", [`mode=${node.mode}`, `literal=${jsonString(node.literal)}`]),
    visitEmphasis: (node: Emphasis) => record("Emphasis", [], node.content.length),
    visitStrong: (node: Strong) => record("Strong", [], node.content.length),
    visitStrikethrough: (node: Strikethrough) => record("Strikethrough", [], node.content.length),
    visitLink: (node: Link) =>
        record(
            "Link",
            [`destination=${optionalString(node.destination)}`, `title=${optionalString(node.title)}`],
            node.content.length
        ),
    visitImage: (node: Image) =>
        record(
            "Image",
            [`source=${optionalString(node.source)}`, `title=${optionalString(node.title)}`],
            node.content.length
        ),
    visitDirective: (node: Directive) =>
        record(
            "Directive",
            directiveFields(node.mode, node.name, node.attributes, node.label?.length ?? null),
            node.label?.length ?? 0
        ),
    visitFootnoteReference: (node: FootnoteReference) => record("FootnoteReference", [`id=${jsonString(node.label)}`])
};

function record(kind: string, fields: readonly string[] = [], children = 0): PendingRecord {
    const fieldText = fields.length === 0 ? "" : ` ${fields.join(" ")}`;
    return { line: (scope) => `${kind} ${scope}${fieldText} children=${children}`, children };
}

function directiveFields(
    mode: "embedded" | "standalone",
    name: string,
    attributes: string | null,
    labelCount: number | null
): readonly string[] {
    return [
        `mode=${mode}`,
        `name=${jsonString(name)}`,
        `attributes=${optionalString(attributes)}`,
        `label=${labelCount ?? "null"}`
    ];
}

function rebased(value: Scope, offset: number): string {
    const start = value.start.line > 0 ? value.start.line - offset : value.start.line;
    const end = value.end.line > 0 ? value.end.line - offset : value.end.line;
    return `scope=${start}:${value.start.column}..${end}:${value.end.column}`;
}

function optionalString(value: string | null): string {
    return value === null ? "null" : jsonString(value);
}

function jsonString(value: string): string {
    return JSON.stringify(value);
}

File: packages/kotlin-markdown-core/src/commonMain/kotlin/com/nouprax/markdown/core/walker/MarkupDumper.kt
@file:kotlin.jvm.JvmName("MarkdownCoreKt")
@file:kotlin.jvm.JvmMultifileClass

package com.nouprax.markdown.core

/** Produces the canonical diagnostic tree for immutable Markdown markup. */
public object MarkupDumper {
    /** Returns the canonical diagnostic dump for [document] and its
     * descendants, resolving absolute scopes through the snapshot. */
    public fun dump(document: Document): String = dump(document, document)

    /**
     * Returns the canonical diagnostic dump for the subtree rooted at
     * [node]. Scopes print with the subtree as origin: the root's start
     * line becomes line 1, later lines shift by the same amount, and
     * columns are line-local and unchanged. Position-free markers
     * (`0:0..0:0`) print unchanged.
     */
    public fun dump(
        document: Document,
        node: Markup,
    ): String {
        val origin = node.scope.start.line
        val offset = if (origin > 0) origin - 1 else 0
        val visitor = DumpVisitor()
        val remainingChildren = mutableListOf<Int>()
        val lines = mutableListOf<String>()

        MarkupWalker.walk(document, node) { event, current, scope ->
            when (event) {
                WalkEvent.ENTERING -> {
                    val record = current.accept(visitor)
                    val line = record.line(scope, offset)
                    if (remainingChildren.isEmpty()) {
                        lines += line
                    } else {
                        val parent = remainingChildren.lastIndex
                        val prefix =
                            remainingChildren
                                .dropLast(1)
                                .joinToString(separator = "") { if (it > 0) "│   " else "    " }
                        val connector = if (remainingChildren[parent] == 1) "└── " else "├── "
                        lines += prefix + connector + line
                        remainingChildren[parent] -= 1
                    }
                    remainingChildren += record.children
                }

                WalkEvent.EXITING -> {
                    check(remainingChildren.removeAt(remainingChildren.lastIndex) == 0)
                }
            }
        }
        return lines.joinToString(separator = "\n", postfix = "\n")
    }
}

private data class DumpRecord(
    val kind: String,
    val fields: kotlin.collections.List<String>,
    val children: Int,
) {
    fun line(
        scope: Scope,
        offset: Int,
    ): String {
        val fieldText = if (fields.isEmpty()) "" else " ${fields.joinToString(" ")}"
        return "$kind ${scopeText(scope, offset)}$fieldText children=$children"
    }
}

private class DumpVisitor : MarkupVisitor<DumpRecord> {
    override fun visit(node: Document): DumpRecord = record("Document", children = node.content.size)

    override fun visit(node: BlockQuote): DumpRecord = record("BlockQuote", children = node.content.size)

    override fun visit(node: Paragraph): DumpRecord = record("Paragraph", children = node.content.size)

    override fun visit(node: Heading): DumpRecord =
        record("Heading", fields = listOf("level=${node.level}"), children = node.content.size)

    override fun visit(node: ThematicBreak): DumpRecord = record("ThematicBreak")

    override fun visit(node: List): DumpRecord =
        record(
            "List",
            fields =
                listOf(
                    "flavor=${node.flavor.token()}",
                    "start=${node.start ?: "null"}",
                    "tight=${node.tight}",
                ),
            children = node.items.size,
        )

    override fun visit(node: ListItem): DumpRecord =
        record(
            "ListItem",
            fields = listOf("checked=${node.checked ?: "null"}"),
            children = node.content.size,
        )

    override fun visit(node: CodeBlock): DumpRecord =
        record(
            "CodeBlock",
            fields =
                listOf(
                    "mode=${node.mode.token()}",
                    "info=${optionalString(node.info)}",
                    "language=${optionalString(node.language)}",
                    "literal=${jsonString(node.literal)}",
                    "fenced=${node.fenced}",
                    "closed=${node.closed}",
                ),
        )

    override fun visit(node: HTMLBlock): DumpRecord =
        record("HTMLBlock", fields = listOf("comment=${node.comment}", "literal=${jsonString(node.literal)}"))

    override fun visit(node: FormulaBlock): DumpRecord =
        record(
            "FormulaBlock",
            fields = listOf("mode=${node.mode.token()}", "literal=${jsonString(node.literal)}"),
        )

    override fun visit(node: Table): DumpRecord =
        record(
            "Table",
            fields = listOf("alignments=[${node.alignments.joinToString(",") { it.token() }}]"),
            children = 1 + node.rows.size,
        )

    override fun visit(node: TableRow): DumpRecord =
        record(
            "TableRow",
            fields = listOf("isHeader=${node.isHeader}"),
            children = node.cells.size,
        )

    override fun visit(node: TableCell): DumpRecord = record("TableCell", children = node.content.size)

    override fun visit(node: DirectiveBlock): DumpRecord =
        record(
            "DirectiveBlock",
            fields = directiveFields(node.mode, node.name, node.attributes),
            children = (if (node.label == null) 0 else 1) + node.content.size,
        )

    override fun visit(node: DirectiveLabel): DumpRecord = record("DirectiveLabel", children = node.content.size)

    override fun visit(node: FootnoteDefinition): DumpRecord =
        record(
            "FootnoteDefinition",
            fields = listOf("id=${jsonString(node.label)}"),
            children = node.content.size,
        )

    override fun visit(node: Text): DumpRecord = record("Text", fields = listOf("literal=${jsonString(node.literal)}"))

    override fun visit(node: SoftBreak): DumpRecord = record("SoftBreak")

    override fun visit(node: LineBreak): DumpRecord = record("LineBreak")

    override fun visit(node: Code): DumpRecord =
        record(
            "Code",
            fields = listOf("mode=${node.mode.token()}", "literal=${jsonString(node.literal)}"),
        )

    override fun visit(node: HTML): DumpRecord =
        record("HTML", fields = listOf("comment=${node.comment}", "literal=${jsonString(node.literal)}"))

    override fun visit(node: Formula): DumpRecord =
        record(
            "Formula",
            fields = listOf("mode=${node.mode.token()}", "literal=${jsonString(node.literal)}"),
        )

    override fun visit(node: Emphasis): DumpRecord = record("Emphasis", children = node.content.size)

    override fun visit(node: Strong): DumpRecord = record("Strong", children = node.content.size)

    override fun visit(node: Strikethrough): DumpRecord = record("Strikethrough", children = node.content.size)

    override fun visit(node: Link): DumpRecord =
        record(
            "Link",
            fields =
                listOf(
                    "destination=${optionalString(node.destination)}",
                    "title=${optionalString(node.title)}",
                ),
            children = node.content.size,
        )

    override fun visit(node: Image): DumpRecord =
        record(
            "Image",
            fields =
                listOf(
                    "source=${optionalString(node.source)}",
                    "title=${optionalString(node.title)}",
                ),
            children = node.content.size,
        )

    override fun visit(node: Directive): DumpRecord =
        record(
            "Directive",
            fields = directiveFields(node.mode, node.name, node.attributes),
            children = if (node.label == null) 0 else 1,
        )

    override fun visit(node: FootnoteReference): DumpRecord =
        record("FootnoteReference", fields = listOf("id=${jsonString(node.label)}"))

    override fun visit(node: ReferenceDefinition): DumpRecord =
        record(
            "ReferenceDefinition",
            fields =
                listOf(
                    "label=${jsonString(node.label)}",
                    "destination=${optionalString(node.destination)}",
                    "title=${optionalString(node.title)}",
                ),
        )

    override fun visit(node: LinkReference): DumpRecord =
        referenceRecord("LinkReference", node.label, node.form, node.content.size)

    override fun visit(node: ImageReference): DumpRecord =
        referenceRecord("ImageReference", node.label, node.form, node.content.size)

    override fun visit(node: CrossLink): DumpRecord =
        record("CrossLink", fields = listOf("reference=${jsonString(node.reference)}"))

    override fun visit(node: Embed): DumpRecord =
        record("Embed", fields = listOf("reference=${jsonString(node.reference)}"))
}

private fun referenceRecord(
    kind: String,
    label: String,
    form: ReferenceForm,
    children: Int,
): DumpRecord = DumpRecord(kind, listOf("label=${jsonString(label)}", "form=${form.token()}"), children)

private fun ReferenceForm.token(): String =
    when (this) {
        ReferenceForm.FULL -> "full"
        ReferenceForm.COLLAPSED -> "collapsed"
        ReferenceForm.SHORTCUT -> "shortcut"
    }

private fun record(
    kind: String,
    fields: kotlin.collections.List<String> = emptyList(),
    children: Int = 0,
): DumpRecord = DumpRecord(kind, fields, children)

private fun directiveFields(
    mode: PlacementMode,
    name: String,
    attributes: String?,
): kotlin.collections.List<String> =
    listOf(
        "mode=${mode.token()}",
        "name=${jsonString(name)}",
        "attributes=${optionalString(attributes)}",
    )

private fun scopeText(
    value: Scope,
    offset: Int,
): String {
    val start = if (value.start.line > 0) value.start.line - offset else value.start.line
    val end = if (value.end.line > 0) value.end.line - offset else value.end.line
    return "scope=$start:${value.start.column}..$end:${value.end.column}"
}

private fun optionalString(value: String?): String = value?.let(::jsonString) ?: "null"

private fun PlacementMode.token(): String = name.lowercase()

private fun ListFlavor.token(): String = name.lowercase()

private fun TableAlignment.token(): String = name.lowercase()

private fun jsonString(value: String): String =
    buildString {
        append('"')
        value.forEach { character ->
            when (character) {
                '"' -> {
                    append("\\\"")
                }

                '\\' -> {
                    append("\\\\")
                }

                '\b' -> {
                    append("\\b")
                }

                '\u000c' -> {
                    append("\\f")
                }

                '\n' -> {
                    append("\\n")
                }

                '\r' -> {
                    append("\\r")
                }

                '\t' -> {
                    append("\\t")
                }

                else -> {
                    if (character.code < 0x20) {
                        append("\\u")
                        append(character.code.toString(16).padStart(4, '0'))
                    } else {
                        append(character)
                    }
                }
            }
        }
        append('"')
    }
